[Site design revamp] Add helper text to the bottom of the theme picker

mkevins opened the pull request:
Fixes #16478 

### Description

This PR adds a bottom helper text to the theme picker screen by implementing an optional footer parameter to layouts recycle view. This is useful for working around scroll issues in the layout that houses the recycler view. It introduces a new abstraction `LayoutRowViewHolder` from which `LayoutItemViewHolder` and `LayoutFooterViewHolder` extend:

```mermaid
classDiagram
  class LayoutRowViewHolder
  <<Sealed>> LayoutRowViewHolder
  LayoutRowViewHolder <|-- LayoutItemViewHolder: Extends
  LayoutRowViewHolder <|-- LayoutFooterViewHolder: Extends
```

The footer is inflated when needed from the passed resource id (the overriden item count is incremented when not null).

It also adds the footer as a `LayoutCategoryViewType` so that it isn't mixed with the regular items during view recycling.

## To test:

### Theme picker

1. Start the site creation flow
2. Select an intent (or skip)
3. Type a site name (or skip)
4. Expect to see the theme picker screen
5. Scroll to the bottom of the theme picker screen
6. Expect to see the helper text in the footer at the bottom of the screen

### Page picker

1. Start the page picker flow (create a new page from the page list screen)
2. Expect to see the page picker screen
3. Scroll to the bottom of the page layout categories
4. Expect that the footer is absent

## Regression Notes
1. Potential unintended areas of impact
Page layout picker

7. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated testing.

8. What automated tests I added (or what prevented me from doing so)
Our tests are currently structured to test the view models, but these changes don't rely on any view model logic.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mkevins committed:
Add bottom helper text to theme picker

This implements an optional footer parameter to layouts recycle view,
useful for working around scroll issues in the screen where it is used.
It introduces a new abstraction `LayoutRowViewHolder` from which
`LayoutItemViewHolder` and `LayoutFooterViewHolder` extend.

The footer is inflated when needed from the passed resource id (the
overriden item count is incremented when not null).

It also adds the footer as a `LayoutCategoryViewType` so that it isn't
mixed with the regular items during view recycling.

mkevins committed:
Merge branch 'feature/site-design-revamp' into feature/site-design-revamp--add-helper-text-to-bottom-as-list-item

mkevins committed:
Fix detekt issues

mkevins committed:
Fix lint

ovitrif reviewed (commented):
Great work @mkevins!

The solution is elegant imho and works as expected 👍 

I did notice two small issues:

1. Sometimes the `Choose a theme` screen starts scrolled to the bottom of the recycler view (this happens only when the device is proxied to use a wpcom sandbox with the pending api changes active):

https://user-images.githubusercontent.com/4588074/168563084-bf76e10d-913f-45f8-9e32-1c41ef96bd35.mp4

2. The style of the footer text might need a few adjustments to match the i3 design:

| Expected | Actual |
| --- | --- |
| <img width="383" alt="expected" src="https://user-images.githubusercontent.com/4588074/168563526-251fdc24-1f78-4559-8843-b89be269432c.png"> | <img width="383" alt="actual" src="https://user-images.githubusercontent.com/4588074/168563788-2bdfd363-928c-467f-8b41-cfa58a19dd03.png"> |

I was thinking that achieving this might be easier with reusing an existing style after we merge #16556.

mkevins committed:
Merge branch 'feature/site-design-revamp' into feature/site-design-revamp--add-helper-text-to-bottom-as-list-item

mkevins committed:
Make some style adjustments

mkevins committed:
Only show footer when list is populated

This prevents an issue where the fetched categories arriving later after
the layout caused the list to be scrolled to the bottom (maintaining a
view of the footer as the new items were added).

mkevins commented:
Thanks for reviewing and testing Ovi, and for the great feedback!

> I was thinking that achieving this might be easier with reusing an existing style after we merge #16556.

Spot on. I was hoping to consolidate some of this to be less ad-hoc, and the PR mentioned allows that. I updated to use that, and adjusted a few other things (https://github.com/wordpress-mobile/WordPress-Android/pull/16555/commits/01d5a852b2de2dba5ccb37d7599e10d01acdce9a) (e.g. the icon was dark in dark mode, and that should be fixed now).

> Sometimes the Choose a theme screen starts scrolled to the bottom of the recycler view (this happens only when the device is proxied to use a wpcom sandbox with the pending api changes active):

I observed this too, even without sandboxing. I believe what is happening is that the footer is displayed without any items, and when the fetch response comes back, the items are added, but the recycler view maintains the scroll position to be aligned with the footer. I pushed a change (https://github.com/wordpress-mobile/WordPress-Android/pull/16555/commits/381266508e5f0e1b655eeb74b7bdcc4a48e38682) which should fix this by conditionally displaying the footer, only when there are items.

This is ready for another pass. :+1:

ovitrif commented:
> Thanks for reviewing and testing Ovi, and for the great feedback!
> 
> > I was thinking that achieving this might be easier with reusing an existing style after we merge #16556.
> 
> Spot on. I was hoping to consolidate some of this to be less ad-hoc, and the PR mentioned allows that. I updated to use that, and adjusted a few other things ([01d5a85](https://github.com/wordpress-mobile/WordPress-Android/commit/01d5a852b2de2dba5ccb37d7599e10d01acdce9a)) (e.g. the icon was dark in dark mode, and that should be fixed now).
> 
> > Sometimes the Choose a theme screen starts scrolled to the bottom of the recycler view (this happens only when the device is proxied to use a wpcom sandbox with the pending api changes active):
> 
> I observed this too, even without sandboxing. I believe what is happening is that the footer is displayed without any items, and when the fetch response comes back, the items are added, but the recycler view maintains the scroll position to be aligned with the footer. I pushed a change ([3812665](https://github.com/wordpress-mobile/WordPress-Android/commit/381266508e5f0e1b655eeb74b7bdcc4a48e38682)) which should fix this by conditionally displaying the footer, only when there are items.
> 
> This is ready for another pass. 👍

I gave this another pass and I feel it's ready to be merged:
- the issue when the list was scrolled to the bottom (with the helper text visible) didn't reproduce
- the design looks good and conforms to the design
- Different variations -- dark mode, tablet, landscape -- look nice

Thank you for your work on this @mkevins 🙇
